common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+ */
+class FileDescriptor {
+public:
+    static constexpr int invalid_descriptor = -1;
+
+    virtual ~FileDescriptor() = default;
+    virtual int fd()          = 0;
+    virtual bool close()      = 0;
+};
+
+/**
+ * File descriptor implementation.
+ * This class is basically a wrapper around an `int fd`
+ * This class will be aggregated from all classes modeling OS resources implementing the file
+ * descriptor interface.
+ */
+class FileDescriptorImpl : public FileDescriptor {
+public:
+    explicit FileDescriptorImpl(int fd) : m_fd(fd)
+    {
+        // TODO: error if FileDescriptor::invalid_descriptor == fd;
+    }
+
+    /**
+     * Delete copy constructor to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl(const FileDescriptorImpl &) = delete;
+
+    /**
+     * Delete assignment operator to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl &operator=(const FileDescriptorImpl &) = delete;
+
+    /**
+     * A move constructor allows the resources owned by an rvalue object to be moved into an
+     * lvalue without creating its copy.
+     */
+    FileDescriptorImpl(FileDescriptorImpl &&obj)
+    {
+        this->m_fd = obj.m_fd;
+        obj.m_fd   = FileDescriptor::invalid_descriptor;
+    }
+
+    /**
+     * Close socket on destructor (if it is still open)
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used:
+     * - in select(), poll() or epoll() to wait for events on this descriptor.
+     * - in send() family of functions to send data through the socket connection.
+     * - in recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());

```

Comment:
The same, simply have pure virtual const and non-const interfaces.
```suggestion
    virtual socklen_t &length() = 0;
```
BTW -

> <sys/socket.h> makes available a type, socklen_t, which is an unsigned opaque integral type of length of at least 32 bits.

So why do you return a reference to it, adding the complexity of const/non-const, life cycle of the object in question, etc? Simply return a _copy_ of it.

Reply:
The intention of having a reference was to use it in `::recvfrom()`. 
However, as @arnout has pointed out [here](https://github.com/prplfoundation/prplMesh/pull/1499/files#r449660712), it's not right.